A/B Testing via LaunchDarkly (#139212)
elastic/kibana: x-pack/plugins/cloud_integrations/cloud_experiments/README.mdx

---
id: kibCloudExperimentsPlugin
slug: /kibana-dev-docs/key-concepts/cloud-experiments-plugin
title: Cloud Experiments service
description: The Cloud Experiments Service provides the necessary APIs to implement A/B testing scenarios, fetching the variations in configuration and reporting back metrics to track conversion rates of the experiments.
date: 2022-09-07
tags: ['kibana', 'dev', 'contributor', 'api docs', 'cloud', 'a/b testing', 'experiments']
---

# Kibana Cloud Experiments Service

> The Cloud Experiments Service provides the necessary APIs to implement A/B testing scenarios, fetching the variations in configuration and reporting back metrics to track conversion rates of the experiments.

Reviewer: nit: Maybe worth being super clear this is only active on cloud? (in case it isn't obvious from the name)

```suggestion
The Cloud Experiments Service provides the necessary APIs to implement A/B testing scenarios, fetching the variations in configuration and reporting back metrics to track conversion rates of the experiments.

The `cloudExperiments` plugin is disabled by default and only enabled on Elastic Cloud deployments.
```

## Public API

If you are developing a feature that needs to use a feature flag, or you are implementing an A/B-testing scenario, this is how you should fetch the value of your feature flags (for either server and browser side code):

> First, you should declare the optional dependency on this plugin (bear in mind that the plugin is disabled by default, so adding this plugin to the `requiredPlugins` will disable your plugin when this one is disabled).

Reviewer: ```suggestion
First, you should declare the optional dependency on this plugin. Do not list it in your `requiredPlugins`, as this plugin is disabled by default and only enabled in Cloud deployments. Adding it to your `requiredPlugins` will cause Kibana to refuse to start by default.
```

```json
// plugin/kibana.json
{
  "id": "myPlugin",
  "optionalPlugins": ["cloudExperiments"]
}
```

> Please, be aware that your plugin may run even when the `cloudExperiment` plugin is disabled. Make sure to declare it as an optional dependency in your plugin's Typescript contract to remind you that it might not always be available.

Reviewer: ```suggestion
Please, be aware that your plugin will run even when the `cloudExperiment` plugin is disabled. Make sure to declare it as an optional dependency in your plugin's TypeScript contract to remind you that it might not always be available.
```

### Fetching the value of the feature flags

First, make sure that your feature flag is listed in [`FEATURE_FLAG_NAMES`](./common/constants.ts). 
Then, you can fetch the value of your feature flag by using the API `cloudExperiments.getVariation` as follows:

```ts
import type { CoreSetup, CoreStart, Plugin } from '@kbn/core/(public|server)';
import type { 
  CloudExperimentsPluginSetup, 
  CloudExperimentsPluginStart
} from '@kbn/cloud-experiments-plugin/common';

interface SetupDeps {
  cloudExperiments?: CloudExperimentsPluginSetup;
}

interface StartDeps {
  cloudExperiments?: CloudExperimentsPluginStart;
}

export class MyPlugin implements Plugin<void, void, SetupDeps, StartDeps> {
  public setup(core: CoreSetup, deps: SetupDeps) {

>     this.doSomethingBasedOnFeatureFlag(deps.cloudExperiments);

Reviewer: Just want to note that if we decide to leave `getVariation` out of the `setup` lifecycle (see my other comment), we should probably remove this part of the example.
